src/pg_query_parse_summary.c

```diff
@@ -0,0 +1,832 @@
+#include "pg_query_outfuncs.h"
+
+#include "parser/parser.h"
+#include "parser/scanner.h"
+#include "parser/scansup.h"
+
+#include "nodes/nodeFuncs.h"
+#include "utils/builtins.h"
+
+#include "protobuf/pg_query.pb-c.h"
+
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include "pg_query_summary.h"
+
+// NOTE regarding cte_names and from_clause_items:
+// 1. We process cte_names as we iterate through the tree.
+// 2. We only add items to tables if they are *not* in cte_names.
+// 3. CTEs can be referenced *after* table names.
+// 4. Because of this, we have to store from_clause_items and process it after the initial tree walk.
+//
+// If we try to do it entirely on the initial pass, we wind up CTEs being
+// included in state->tables, and thus in summary->tables.
+
+// We need to declare this early because it gets called before it's defined.
+void add_table(WalkState* state, enum ContextType context, char* schemaname, char* relname);
+
+// small utility function for creating error messages.
+PgQueryError* pg_query_parse_summary_mkerror(const char* message)
+{
+	PgQueryError* error = palloc(sizeof(PgQueryError));
+	error->message = strdup(message);
+	return error;
+}
+
+// Given a Node (node) and WalkState (state), wraps node in a NodeWithContext
+// and adds it to the state->from_clause_items list.
+void add_from_clause_item(Node *node, enum ContextType context, WalkState *state) {
+	if (!node)
+		return;
+
+	NodeWithContext* nwc = palloc(sizeof(NodeWithContext));
+	nwc->node = node;
+	nwc->context = context;
+
+	state->from_clause_items = lappend(state->from_clause_items, nwc);
+}
+
+// Given a List*-of-Nodes (clause) and WalkState (state), wrap them all
+// in NodeWithContext structs and add them to the state->from_clause_items list.
+void add_from_clause_items(List *clause, enum ContextType context, WalkState *state) {
+	if (!clause)
+		return;
+
+	ListCell *lc = NULL;
+	foreach(lc, clause) {
+		Node* node = lfirst(lc);
+		add_from_clause_item(node, context, state);
+	}
+}
+
+void add_function(List* funcname, enum ContextType context, WalkState* state){
+	StringInfoData buf;
+	initStringInfo(&buf);
+
+	bool first = true;
+	ListCell* lc = NULL;
+	foreach(lc, funcname) {
+		// Add a dot separator before everything but the first item.
+		if (first)
+			first = false;
+		else
+			appendStringInfoChar(&buf, '.');
+
+		String* str = (String*)lfirst(lc);
+		appendStringInfoString(&buf, quote_identifier(str->sval));
+	}
+
+	char* name = buf.data;
+	String* str_name = makeString(name);
+	Function* func = palloc(sizeof(Function));
+
+	func->name = str_name->sval;
+
+	// TODO: Verify this definition of schema_name is correct
+	if (list_length(funcname) > 1) {
+		String* first = linitial(funcname);
+		func->schema_name = first->sval;
+	} else {
+		func->schema_name = NULL;
+	}
+
+	// TODO: Verify this definition of function_name is correct.
+	String* last = llast(funcname);
+	func->function_name = last->sval;
```

Comment:
This is correct, but I'd keep it shorter without the extra variable:

```suggestion
	func->function_name = llast(funcname)->sval;
```